app/src/main/java/nl/eduvpn/app/service/ConnectionService.kt

```diff
@@ -0,0 +1,235 @@
+/*
+ *  This file is part of eduVPN.
+ *
+ *     eduVPN is free software: you can redistribute it and/or modify
+ *     it under the terms of the GNU General Public License as published by
+ *     the Free Software Foundation, either version 3 of the License, or
+ *     (at your option) any later version.
+ *
+ *     eduVPN is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ *     You should have received a copy of the GNU General Public License
+ *     along with eduVPN.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package nl.eduvpn.app.service
+
+import android.app.Activity
+import android.app.PendingIntent
+import android.content.Intent
+import android.net.Uri
+import android.os.Parcelable
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import net.openid.appauth.*
+import net.openid.appauth.browser.BrowserDenyList
+import net.openid.appauth.browser.VersionedBrowserMatcher
+import nl.eduvpn.app.BuildConfig
+import nl.eduvpn.app.MainActivity
+import nl.eduvpn.app.R
+import nl.eduvpn.app.entity.DiscoveredAPI
+import nl.eduvpn.app.entity.Instance
+import nl.eduvpn.app.utils.ErrorDialog.show
+import nl.eduvpn.app.utils.Log
+import nl.eduvpn.app.utils.runCatchingCoroutine
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+/**
+ * The connection service takes care of building up the URLs and validating the result.
+ * Created by Daniel Zolnai on 2016-10-11.
+ *
+ * @param preferencesService The preferences service used to store temporary data.
+ * @param historyService     History service for storing data for long-term usage
+ * @param securityService    For security related tasks.
+ */
+class ConnectionService(private val preferencesService: PreferencesService,
+                        private val historyService: HistoryService,
+                        private val securityService: SecurityService) {
+
+    private var authorizationService: AuthorizationService? = null
+
+    fun onStart(activity: Activity?) {
```

Comment:
We should make activity a non-null type, otherwise it will crash a few lines later

Reply:
Fixed in https://github.com/eduvpn/android/pull/315/commits/c240b678145953049e7ff9227caf4d1e605e8958